### ISSUE TITLE
Gy Redirect Policy Rule removed after Validity Timer triggered CCR-U with FUA Redirect

### ISSUE THREAD

**Issue by jpad-freedomfi** [magma/magma#6993]
### Your Environment

- **Version:** 1.4 orc8r + FeG, AGW Bare Metal 1.4.0-1618930512-8dcd7a4e
- **Affected Component:** FeG + Access Gateway
- **Affected Subcomponent:** Sessiond / Session Proxy
- **Deployment Environment:** e.g. AWS (orc8r), bare-metal (AGW)

### Describe the Issue

We are regression testing 1.4 as we upgrade from 1.3. We are testing UE attach on SIM with zero balance in OCS. We expect that CCR-I is replied with FUI/FUA redirect. We have redirectd enabled on the gateway. 

On attach, UE is connected, has zero balance, and redirect works. When validity time expires, the CCA-U has FUA redirect and policy rule for redirect is deactivated at which point UE can consume traffic and it isn't accounted as it still has no quota. 

**To Reproduce**
1. Setup feg_lte network with FeG Gy + Gx relay
2. Provision a SIM in OCS with zero balance
3. Attach UE to network
4. Observe CCA-I contains FUI/FUA redirect
5. try to surf to http://wumo.com/wumo and see that redirect works
6. leave UE idle until validity timer expires and triggers CCR-U 
7. observe CCA-U contains FUI/FUA redirect
8. Observe in sessiond logs that redirect policy rule gets deactivated and leaves open rule "rg1_high_speed" in place
9. try to surf any website and see that data access is possible. 

**Expected behavior**
A CCA-U with FUA redirect should add or keep the redirect policy rule in place, it should not remove the redirect policy rule regardless of GSU value. 

**Key Log Snippet**
```
May 18 06:33:02 agw sessiond[2807]: I0518 06:33:02.710675  2807 SessionState.cpp:497] Updating used charging credit for Rule=rg1_high_speed Rating Group=1 Service Identifier=0
May 18 06:33:12 agw sessiond[2807]: I0518 06:33:12.709408  2824 GrpcMagmaUtils.cpp:38]
May 18 06:33:12 agw sessiond[2807]:   magma.lte.RuleRecordTable {
May 18 06:33:12 agw sessiond[2807]:       records {
May 18 06:33:12 agw sessiond[2807]:         sid: "IMSI315010999900108"
May 18 06:33:12 agw sessiond[2807]:         rule_id: "internal_default_drop_flow_rule"
May 18 06:33:12 agw sessiond[2807]:         ue_ipv4: "192.168.128.14"
May 18 06:33:12 agw sessiond[2807]:       }
May 18 06:33:12 agw sessiond[2807]:       records {
May 18 06:33:12 agw sessiond[2807]:         sid: "IMSI315010999900108"
May 18 06:33:12 agw sessiond[2807]:         rule_id: "rg1_high_speed"
May 18 06:33:12 agw sessiond[2807]:         ue_ipv4: "192.168.128.14"
May 18 06:33:12 agw sessiond[2807]:       }
May 18 06:33:12 agw sessiond[2807]:       records {
May 18 06:33:12 agw sessiond[2807]:         sid: "IMSI315010999900108"
May 18 06:33:12 agw sessiond[2807]:         rule_id: "redirect"
May 18 06:33:12 agw sessiond[2807]:         ue_ipv4: "192.168.128.14"
May 18 06:33:12 agw sessiond[2807]:       }
May 18 06:33:12 agw sessiond[2807]:       epoch: 1621318255
May 18 06:33:12 agw sessiond[2807]:   }
May 18 06:33:12 agw sessiond[2807]: I0518 06:33:12.710125  2807 SessionState.cpp:497] Updating used charging credit for Rule=rg1_high_speed Rating Group=1 Service Identifier=0
May 18 06:33:12 agw sessiond[2807]: I0518 06:33:12.710211  2807 LocalSessionManagerHandler.cpp:87] Sending 1 charging updates and 0 monitor updates to OCS and PCRF
May 18 06:33:12 agw sessiond[2807]: I0518 06:33:12.712232  2807 GrpcMagmaUtils.cpp:38]
May 18 06:33:12 agw sessiond[2807]:   magma.lte.UpdateSessionRequest {
May 18 06:33:12 agw sessiond[2807]:       updates {
May 18 06:33:12 agw sessiond[2807]:         usage {
May 18 06:33:12 agw sessiond[2807]:           bytes_tx: 632
May 18 06:33:12 agw sessiond[2807]:           bytes_rx: 3234
May 18 06:33:12 agw sessiond[2807]:           charging_key: 1
May 18 06:33:12 agw sessiond[2807]:           type: VALIDITY_TIMER_EXPIRED
May 18 06:33:12 agw sessiond[2807]:           requested_units {
May 18 06:33:12 agw sessiond[2807]:           }
May 18 06:33:12 agw sessiond[2807]:         }
May 18 06:33:12 agw sessiond[2807]:         session_id: "IMSI315010999900108-399779"
May 18 06:33:12 agw sessiond[2807]:         request_number: 1
May 18 06:33:12 agw sessiond[2807]:         sid: "IMSI315010999900108"
May 18 06:33:12 agw sessiond[2807]:         ue_ipv4: "192.168.128.14"
May 18 06:33:12 agw sessiond[2807]:         spgw_ipv4: "10.0.2.1"
May 18 06:33:12 agw sessiond[2807]:         apn: "internet"
May 18 06:33:12 agw sessiond[2807]:         plmn_id: "315010"
May 18 06:33:12 agw sessiond[2807]:         imsi_plmn_id: "315010"
May 18 06:33:12 agw sessiond[2807]:         user_location: "\202\023\005\020\000\001\023\005\020\000\006\"\001"
May 18 06:33:12 agw sessiond[2807]:         tgpp_ctx {
May 18 06:33:12 agw sessiond[2807]:           gx_dest_host: "gx00.i-new.com"
May 18 06:33:12 agw sessiond[2807]:           gy_dest_host: "gy00.inew-cs.com"
May 18 06:33:12 agw sessiond[2807]:         }
May 18 06:33:12 agw sessiond[2807]:         common_context {
May 18 06:33:12 agw sessiond[2807]:           sid {
May 18 06:33:12 agw sessiond[2807]:             id: "IMSI315010999900108"
May 18 06:33:12 agw sessiond[2807]:           }
May 18 06:33:12 agw sessiond[2807]:           ue_ipv4: "192.168.128.14"
May 18 06:33:12 agw sessiond[2807]:           apn: "internet"
May 18 06:33:12 agw sessiond[2807]:           teids {
May 18 06:33:12 agw sessiond[2807]:             enb_teid: 1532801
May 18 06:33:12 agw sessiond[2807]:             agw_teid: 4
May 18 06:33:12 agw sessiond[2807]:           }
May 18 06:33:12 agw sessiond[2807]:         }
May 18 06:33:12 agw sessiond[2807]:       }
May 18 06:33:12 agw sessiond[2807]:   }
May 18 06:33:14 agw sessiond[2807]: I0518 06:33:14.154769  2807 GrpcMagmaUtils.cpp:38]
May 18 06:33:14 agw sessiond[2807]:   magma.lte.UpdateSessionResponse {
May 18 06:33:14 agw sessiond[2807]:       responses {
May 18 06:33:14 agw sessiond[2807]:         success: true
May 18 06:33:14 agw sessiond[2807]:         sid: "IMSI315010999900108"
May 18 06:33:14 agw sessiond[2807]:         charging_key: 1
May 18 06:33:14 agw sessiond[2807]:         credit {
May 18 06:33:14 agw sessiond[2807]:           validity_time: 900
May 18 06:33:14 agw sessiond[2807]:           is_final: true
May 18 06:33:14 agw sessiond[2807]:           final_action: REDIRECT
May 18 06:33:14 agw sessiond[2807]:           granted_units {
May 18 06:33:14 agw sessiond[2807]:             total {
May 18 06:33:14 agw sessiond[2807]:             }
May 18 06:33:14 agw sessiond[2807]:             tx {
May 18 06:33:14 agw sessiond[2807]:             }
May 18 06:33:14 agw sessiond[2807]:             rx {
May 18 06:33:14 agw sessiond[2807]:             }
May 18 06:33:14 agw sessiond[2807]:           }
May 18 06:33:14 agw sessiond[2807]:           redirect_server {
May 18 06:33:14 agw sessiond[2807]:             redirect_address_type: URL
May 18 06:33:14 agw sessiond[2807]:             redirect_server_address: "http://www.sersimple.com/PortalCautivo.php"
May 18 06:33:14 agw sessiond[2807]:           }
May 18 06:33:14 agw sessiond[2807]:         }
May 18 06:33:14 agw sessiond[2807]:         result_code: 2001
May 18 06:33:14 agw sessiond[2807]:         tgpp_ctx {
May 18 06:33:14 agw sessiond[2807]:           gx_dest_host: "gx00.i-new.com"
May 18 06:33:14 agw sessiond[2807]:           gy_dest_host: "gy00.inew-cs.com"
May 18 06:33:14 agw sessiond[2807]:         }
May 18 06:33:14 agw sessiond[2807]:         session_id: "IMSI315010999900108-399779"
May 18 06:33:14 agw sessiond[2807]:       }
May 18 06:33:14 agw sessiond[2807]:   }
May 18 06:33:14 agw sessiond[2807]: I0518 06:33:14.155704  2807 ChargingGrant.cpp:87] GSU for RG: 1 IMSI315010999900108-399779 is invalid, but accepting it as it has a final unit action or suspended credit
May 18 06:33:14 agw sessiond[2807]: I0518 06:33:14.155721  2807 SessionCredit.cpp:574] Received a GSU with no valid grants, keeping the same type
May 18 06:33:14 agw sessiond[2807]: I0518 06:33:14.155727  2807 SessionCredit.cpp:189] Received the following credit total_volume=0 tx_volume=0 rx_volume=0 grant_tracking_type=TRACKING_UNSET
May 18 06:33:14 agw sessiond[2807]: I0518 06:33:14.155738  2807 ChargingGrant.cpp:346] IMSI315010999900108-399779 received a credit RG: 1 with final action REDIRECT, redirect_server: http://www.sersimple.com/PortalCautivo.php with expiry timer in 900 seconds
May 18 06:33:14 agw sessiond[2807]: I0518 06:33:14.155781  2807 GrpcMagmaUtils.cpp:38]
May 18 06:33:14 agw sessiond[2807]:   magma.lte.DeactivateFlowsRequest {
May 18 06:33:14 agw sessiond[2807]:       sid {
May 18 06:33:14 agw sessiond[2807]:         id: "IMSI315010999900108"
May 18 06:33:14 agw sessiond[2807]:       }
May 18 06:33:14 agw sessiond[2807]:       rule_ids: "redirect"
May 18 06:33:14 agw sessiond[2807]:       request_origin {
May 18 06:33:14 agw sessiond[2807]:         type: GY
May 18 06:33:14 agw sessiond[2807]:       }
May 18 06:33:14 agw sessiond[2807]:       ip_addr: "192.168.128.14"
May 18 06:33:14 agw sessiond[2807]:   }

```
**Additional context**
Gy pcap, session_proxy log, sessiond full log
[Cyan200.zip](https://github.com/magma/magma/files/6504782/Cyan200.zip)

**Comment by jpad-freedomfi** [comment]
Updated to 1.5.0 (AGW, or8r, and Feg) and retested 

Result was the same, initially redirect worked, but after validity timer expiry and CCR-U with second CCA-U with redirect, the redirect rule is deactivated and UE can access internet. 
[redirect_validity_timer_1.5.0_retest.zip](https://github.com/magma/magma/files/6517328/redirect_validity_timer_1.5.0_retest.zip)

**Comment by ekowtaylor** [comment]
@uri200, @themarwhal  for initial comments

**Comment by themarwhal** [comment]
Will take a look. Attached a potential fix still in WIP.. will backport to 1.4/1.5 once it is merged.

**Comment by themarwhal** [comment]
Didn't meant to close the issue. The fix is now merged into master. I'll work on getting it into 1.5 tomorrow.

**Comment by jpad-freedomfi** [comment]
Thanks @themarwhal ! Can you comment on what the fix enables? Will we have a separate rule that Gx will need to install for HE? or will we just add the HE flow to and existing rule used by Gx?

**Comment by themarwhal** [comment]
> Thanks @themarwhal ! Can you comment on what the fix enables? Will we have a separate rule that Gx will need to install for HE? or will we just add the HE flow to and existing rule used by Gx?

Hi @jpad-freedomfi !
This fix will make it so that step 8 in the issue summary does not happen. Instead of the redirect rule being removed, it should remain in place until new grant is given to the rating group.

**Comment by jpad-freedomfi** [comment]
@themarwhal whaa! sorry, I was mixing issues. Sounds good. Thank you.

**Comment by themarwhal** [comment]
Has been backported into 1.5. The fix will be included in the 1.5.1 release.

**Comment by ekowtaylor** [comment]
Newer release available https://github.com/magma/magma/releases/
